Safari sessionStorage partitioning fix

danimoh opened the pull request:
This pull request introduces a `NonPartitionedSessionStorage` class to circumvent the [issues that Safari introduced in 16.1. by partitioning `sessionStorage` between top-level and iframe contexts](https://bugs.webkit.org/show_bug.cgi?id=247565).
Under the hood, now cookies are used if the `sessionStorage` is detected as being partitioned.
For this, cookie handling has been unified in the `CookieJar` and generally been made more resilient and complete.
On top of the `CookieJar`, which handles plain cookies, a `CookieStorage` has been built which can encode arbitrary data of (practically) arbitrary size into cookies, and encrypts them by default to avoid sensitive data getting leaked to the server.
The cookie-stripping Cookie Monster service worker has now also been enabled as an additional measure to avoid cookies leaking to the server.
I did not observe the previous issues we had a few years ago in Safari with multiple service workers in the same tab (in the top-level window and the iframe). This was however on localhost (with wallet and hub built as production, to enable their service workers). Eventually, I would like to test this in a testnet deployment.

The login flow and swap flow have been adapted to now use the `NonPartitionedSessionStorage` instead of the `sessionStorage`.
These are accompanied with small changes to the Hub to forward the potentially created `CookieStorage` encryption keys.
I'll push those changes tomorrow. They'll also require a Keyguard client version update when released.

The `CookieStorage` and `NonPartitionedSessionStorage` have been implemented in a generic fashion, and not too specific to the login and swap use-cases, with the idea in mind to also be able to use them instead of the Safari keys cookie, to lift the limit on maximum accounts the user can be logged into. Also, the concept of the `NonPartitionedSessionStorage` can be extended to longer-lived storage to maybe be used as a replacement in Safari for the partitioned indexedDB, too.

I guess, I should also add some tests for the new libraries.

danimoh committed:
BrowserStack local iOS testing compatibility

On iOS, BrowserStack routes localhost connections for testing of local code
through an actual domain (bs-local.com) because iOS blocks localhost connections
to content that's not sitting on the phone.
This commit adds changes to treat bs-local.com just like localhost, such that
a locally served dev server can be tested with BrowserStack.

danimoh committed:
Refactor all cookie usages to use CookieJar

danimoh committed:
CookieJar: define known cookies as enum

danimoh committed:
CookieJar: add quota checks for max cookie size and count and recover dropped cookies

danimoh committed:
CookieJar: check allowed chars in cookie name and value

danimoh committed:
CookieStorage

Add CookieStorage as storage alternative to sessionStorage, localStorage,
indexedDB, etc.

danimoh committed:
NonPartitionedSessionStorage

An adapter that uses either sessionStorage or CookieStorage depending
on whether sessionStorage is partitioned or not.

danimoh committed:
NonPartitionedSessionStorage: improve sessionStorage partioning detection

danimoh committed:
NonPartitionedSessionStorage: namespace cookies and sessionStorage entries

danimoh committed:
Import: write seed via NonPartitionedSessionStorage to be available in iframe api

danimoh committed:
Swap: write swap setup result via NonPartitionedSessionStorage to be avialable in swap iframe

danimoh committed:
Enable Cookie Monster service worker

Strip cookies from requests via a service worker to avoid that the cookies
are sent to the server, which are only meant to store data locally. The
Cookie Monster service worker had already been several years ago, however
back then it couldn't be enabled, because Safari had issues with multiple
service workers in the same tab in the top-level window and iframes. These
issues do not seem to exist anymore and the service worker is now safe to
be enabled.

sisou commented on client/src/KeyguardClient.ts:
```diff
@@ -148,9 +148,13 @@ export class KeyguardClient {
+    ): Promise<DerivedAddress[]> {
         return this._iframeRequest<DeriveAddressesRequest, DerivedAddress[]>
-            (KeyguardCommand.DERIVE_ADDRESSES, { keyId, paths });
+            (KeyguardCommand.DERIVE_ADDRESSES, { keyId, paths, temporaryCookieStorageEncryptionKey });
```
Very long variable name :D Do we need the `temporary`? Can it be dropped?

mraveux replied:
Could also be `tmpCookieStorageEncryptionKey` 🤓

danimoh replied:
I went with `tmpCookieEncryptionKey`.

sisou commented on src/lib/CookieJar.js:
```diff
@@ -47,42 +137,36 @@ class CookieJar { // eslint-disable-line no-unused-vars
         if (!str) return [];
 
-        const keys = str.split(',');
+        // previous versions used "," as separator which is strictly speaking not allowed in cookie values
```
:bulb:

sisou commented on src/lib/CookieStorage.js:
```diff
@@ -0,0 +1,312 @@
+     */
+    static _encodeMetadata(metadata) {
+        const encryptionInitVectorSize = metadata.encryptionInitVector ? metadata.encryptionInitVector.byteLength : 0;
+        if (metadata.chunkCount > 127
+            || (metadata.isEncrypted && !metadata.encryptionInitVector)
+            || (metadata.encryptionInitVector && encryptionInitVectorSize !== CookieStorage.ENCRYPTION_INIT_VECTOR_SIZE)
+        ) throw new Errors.KeyguardError('Invalid CookieStorage metadata.');
```
Suggestion to put the first condition already on a new line, indented, to match with the other lines.

```suggestion
        if (
            metadata.chunkCount > 127
            || (metadata.isEncrypted && !metadata.encryptionInitVector)
            || (metadata.encryptionInitVector && encryptionInitVectorSize !== CookieStorage.ENCRYPTION_INIT_VECTOR_SIZE)
        ) throw new Errors.KeyguardError('Invalid CookieStorage metadata.');
```

danimoh replied:
Nah, that doesn't look nice :D

sisou commented on src/lib/CookieStorage.js:
```diff
@@ -0,0 +1,312 @@
+        const firstChunk = CookieJar.readCookie(this._getChunkCookieName(name, 0, namespace));
+        if (!firstChunk) return null;
+        // Read enough base64 data to include the longest possible metadata, if the cookie is long enough.
+        return this._decodeMetadata(Nimiq.BufferUtils.fromBase64Url(firstChunk.substring(0, 23)));
```
The `23` is a bit arbitrary here, without explanation. Is it the length of IV_SIZE + 1 in base64 encoding? Can we maybe make this a formula?

danimoh replied:
I actually calculated it wrong, with the iv size as 16 bytes instead of 12 :D
(But higher numbers are not an issue, it was just base64 decoding more data than necessary.)

sisou commented on src/lib/CookieStorage.js:
```diff
@@ -0,0 +1,312 @@
+    static _getChunkCookieName(name, chunk, namespace = CookieJar.Cookie.NAMESPACE_COOKIE_STORAGE) {
+        name = CookieJar.sanitizeCookieName(name, /* encodeInvalidChars */ false);
+        namespace = CookieJar.sanitizeCookieName(namespace, /* encodeInvalidChars */ false);
+        return `${namespace}${name}${chunk !== 0 ? `${chunk.toString(32)}` : ''}`;
```
`Number.prototype.toString()` supports a radix up to 36 according to Chrome DevTools, if you want to save all the space ;)

danimoh replied:
Well, ok, why not.

sisou commented on src/lib/NonPartitionedSessionStorage.js:
```diff
@@ -0,0 +1,190 @@
+                namespace: CookieJar.Cookie.NAMESPACE_NON_PARTITIONED_SESSION_STORAGE,
+            });
+        }
+        sessionStorage.setItem(
```
Extra blank line to show this is the "else" block.
```suggestion
        }
        
        sessionStorage.setItem(
```

sisou commented on src/lib/CookieJar.js:
```diff
@@ -95,3 +179,40 @@ class CookieJar { // eslint-disable-line no-unused-vars
+    LANGUAGE: /** @type {'lang'} */ ('lang'),
+    KEYS: /** @type {'k'} */ ('k'),
+    REMOVE_KEY: /** @type {'removeKey'} */ ('removeKey'),
+    FLAG_SESSION_STORAGE_NON_PARTITIONED: /** @type {'ssnp'} */ ('ssnp'),
```
```suggestion
    FLAG_SESSION_STORAGE_NOT_PARTITIONED: /** @type {'ssnp'} */ ('ssnp'),
```

sisou commented on src/lib/NonPartitionedSessionStorage.js:
```diff
@@ -0,0 +1,190 @@
+// configured to persist sessions.
+NonPartitionedSessionStorage.COOKIE_MAX_AGE = 7 * 60; // 7 minutes; in seconds
+NonPartitionedSessionStorage.TOP_LEVEL_SESSION_STORAGE_ENTRY = 'topLevelEntry';
+NonPartitionedSessionStorage.IFRAME_SESSION_STORAGE_ENTRY = 'iframeEntry';
```
```suggestion
NonPartitionedSessionStorage.TOP_LEVEL_SESSION_STORAGE_ENTRY = '_topLevelEntry';
NonPartitionedSessionStorage.IFRAME_SESSION_STORAGE_ENTRY = '_iframeEntry';
```

sisou reviewed (commented):
Wow, crazy stuff. Well done, but a lot of time spent on generalization :D

danimoh commented:
Thanks for the review.
Yeah, I saw that the stored swap request will require support for cookie storage of arbitrary size, so that led me somewhat down a rabbit hole with the generalization. But now we can use it for the keys cookie too :)

I addressed your comments @sisou.
See here for the diff: https://github.com/nimiq/keyguard/compare/c2dc8334a6978d51e24bc7ec3287d92fe663026e..aa01a3187b3b3f83fa7b68f025994de2c3ca11dc
